### DISCUSSION TITLE
Obtaining more MIDI device details

### DISCUSSION THREAD

**Discussion by albeardsley** [mido/mido#581] [Q&A] [answered]
Hi there,
I appreciate the work you've done with mido, it has been indispensable for my projects.

I was wondering if it was possible through mido to obtain more information about the plugged in MIDI devices.
Primarily, I was interested in determining the length of the keyboard (e.g., a 61 key piano or an 88 key piano).
But also a range of other information would be very useful, such as whether or not it has a pitchwheel, bluetooth capabilities, current local control status, etc.

Any pointers (no pun intended) on where I could start looking would be much appreciated.
Thanks

**Comment by rdoursenaud** [comment, marked as answer]
Hi and sorry for the late reply.

Unfortunately MIDI 1.0 doesn't provide this kind of capability.

The nearest feature would be using the `Device Inquiry` System Exclusive Universal message (See page 40 of the [MIDI 1.0 detailed specification](https://midi.org/midi-1-0-detailed-specification)) but parsing the response - if the target device supports Device Inquiry - would need access to a database with all the devices and features you're looking for.
A project on its own.

MIDI 2.0 should allow this with MIDI-CI but there is almost no device out there and mido doesn't support it as I'm not well versed into it yet.

Hope this helps.

**Reply by rdoursenaud** [reply]
For what it's worth, the Device Inquiry message to query all devices is `F0 7E 7F 06 01 F7`.

If supported, the device should respond with: `F0 7E <device ID> 06 02 mm ff ff dd dd ss ss ss ss F7`.

| Byte(s) | Description |
|---------|-------------|
| `F0 7E <device ID>` | Universal System Exclusive Non-real time header |
| `06` |General Information (sub-ID#1) |
| `02` |Identity Reply (sub-ID#2) |
| `mm` | Manufacturers System Exclusive id code |
| `ff ff` | Device family code (14 bits, LSB first) |
| `dd dd` |Device family member code (14 bits, LSB first) |
| `ss ss ss ss` | Software revision level. Format device specific |
| `F7` | EOX |

*Extracted from the [MIDI 1.0 detailed specification](https://midi.org/midi-1-0-detailed-specification), page 40*.

**Reply by albeardsley** [reply]
Hey no worries. Thanks for your help.
I'll have a look into it.
Cheers
